packages/model/addon/-private/record-state.ts

```diff
@@ -0,0 +1,392 @@
+    // probably notify here but not yet.
+    setter.call(this, v);
+  };
+  dependentKeyCompat(desc);
+  return desc;
+}
+
+/*
+root
+  empty
+    deleted    // hidden from stateName
+    preloaded  // hidden from stateName
+
+  loading
+     empty     // hidden from stateName
+     preloaded // hidden from stateName
+
+  loaded
+    saved
+    updated
+      uncommitted
+      invalid
+      inFlight
+    created
+      uncommitted
+      invalid
+      inFlight
+
+  deleted
+    saved
+      new      // hidden from stateName
+    uncommitted
+    invalid
+    inFlight
+*/
+export default class RecordState {
+  declare store: CoreStore;
+  declare identifier: StableRecordIdentifier;
+  declare record: Model;
+  declare rs: RequestCache;
+
+  declare pendingCount: number;
+  declare fulfilledCount: number;
+  declare rejectedCount: number;
+  declare recordData: RecordData;
+  declare _errorRequests: any[];
+  declare _lastError: any;
+
+  constructor(record: Model) {
+    const { store } = record;
+
+    let id = record._internalModel.identifier;
+
+    this.record = record;
+    this.recordData = record._internalModel._recordData;
+
+    this.pendingCount = 0;
+    this.fulfilledCount = 0;
+    this.rejectedCount = 0;
+
+    this._errorRequests = [];
+    this._lastError = null;
+
+    let requests = store.getRequestStateService();
+    let notifications = store._notificationManager;
+
+    requests.subscribeForRecord(id, (req) => {
+      if (req.type === 'mutation') {
+        switch (req.state) {
+          case 'pending':
+            this.isSaving = true;
+            break;
+          case 'rejected':
+            this.isSaving = false;
+
+            this._lastError = req;
+            if (!(req.response && isInvalidError(req.response.data))) {
+              this._errorRequests.push(req);
+            }
+
+            this.notify('isError');
+            this.notify('adapterError');
+            this.notify('isValid');
+            break;
+          case 'fulfilled':
+            this._errorRequests = [];
+            this._lastError = null;
+            this.isSaving = false;
+            this.notify('isError');
+            this.notify('adapterError');
+            this.notify('isValid');
+            break;
+        }
+      } else {
+        switch (req.state) {
+          case 'pending':
+            this.pendingCount++;
+            this.notify('isLoading');
+            break;
+          case 'rejected':
+            this.pendingCount--;
+            this._lastError = req;
+            if (!(req.response && isInvalidError(req.response.data))) {
+              this._errorRequests.push(req);
+            }
+            this.notify('isLoading');
+            this.notify('isError');
+            this.notify('adapterError');
+            this.notify('isValid');
+            break;
+          case 'fulfilled':
+            this.pendingCount--;
+            this.fulfilledCount++;
+            this.notify('isLoading');
```

Comment:
Can we extract these notifications into a helper method so we don't accidentally forget to call one.